Experimental: Dataplane Contract first pass

kylebrandt opened the pull request:
First steps in adding Go writers and readers to go sdk that conform to https://github.com/grafana/grafana-plugin-sdk-go/tree/main/data/contract_docs

This is all still a bit rough, but it goes into experimental. Looking to merge so can try out the writers with debug datasource.

kylebrandt committed:
wip: exploring func signature for metrics to create consistent schemas

kylebrandt committed:
populate a function

kylebrandt committed:
start some validation

kylebrandt committed:
more validation

kylebrandt committed:
some cleanup; add metricref idea I am not sure about yet

ryantxu commented on experimental/sdata/sdata.go:
```diff
@@ -0,0 +1,226 @@
+		return fmt.Errorf("type %T is not a valid data frame field type", values)
+	}
+
+	valueField := data.NewField(metricName, l, values) // note
```
^^ worth checking the details on this.  I *think* it encourages a pattern where we copy/allocate the same data 3x+

vs the flavors where you know the length upfront and set at each index.  In that case we only allocate once (+ whatever the initial data was)

kylebrandt replied:
You mean the implementation of data.NewField? (Just checking that you didn't mean to comment on AppendMetricValue below)

kylebrandt replied:
Hmm... look at the implementation, not sure why the vector can not be passed as a whole when the vector is created, there is `newInt8Vector`, not not one that takes the type an asserts it... I can perhaps look into that.

ryantxu replied:
exactly -- assuming we have a typed go struct (not interface{}) I think the type assertions are not necessary. And obviously avoiding a 2nd copy will be helpful :)

kylebrandt committed:
Merge remote-tracking branch 'origin/main' into dataSchemaPlay

kylebrandt committed:
start WideSeries, AddMetric

kylebrandt committed:
remove append/insert for now at least, separate out interfaces

kylebrandt committed:
add GetMetricsRefs for Wide

kylebrandt committed:
add WideFrameSeriesAsMultFrameSeries

kylebrandt committed:
add LongSeries GetMetricRefs

kylebrandt committed:
remove AsWide AsMulti, so Refs is the only reader

kylebrandt committed:
move files

kylebrandt committed:
start on numeric

kylebrandt committed:
more numeric work

kylebrandt committed:
cleanup a test

kylebrandt committed:
comments on test

kylebrandt committed:
subdivide that test

torkelo commented:
This looks pretty good!

kylebrandt committed:
Start reworking validate

kylebrandt committed:
wip: ignored fields, interface inconsistent while on this branch

kylebrandt committed:
Add more ignoredFields to GetRefs

kylebrandt committed:
file shuffle

kylebrandt committed:
update series long

kylebrandt committed:
split test files

kylebrandt committed:
TODO: Change wide builder not to be interface

kylebrandt committed:
remove SeriesWriter interface

kylebrandt committed:
remove pointer recv

kylebrandt committed:
notes to self on empty, next todo I think

kylebrandt committed:
start reworking for empty responses

kylebrandt committed:
more work on empty

kylebrandt committed:
More work on empty

kylebrandt committed:
wip: redo validation, starting with multi, not building, may forcepush

kylebrandt committed:
make it build again/some multiframe tests passing

kylebrandt committed:
work on combining validate/getrefs (currently with multi)

kylebrandt committed:
work on combining validate/getrefs (wide, TODO: long)
also todo making quite a mess, some constants or errors types, funcs shorter if possible

kylebrandt committed:
de-dup some of code, start long validate/getrefs combine

kylebrandt committed:
restore long

kylebrandt committed:
Merge remote-tracking branch 'origin/main' into dataSchemaPlay

kylebrandt committed:
test for empty, need to redo NewWide

kylebrandt committed:
redo how NewWide works

kylebrandt committed:
remove comment
If container, slice still needs to a pointer, so no point that I see right now
pun intented

kylebrandt committed:
some comments

kylebrandt committed:
start get reader

kylebrandt committed:
Change wide to underlying multiple frames, TODO long

kylebrandt committed:
Change long to underlying multiple frames

kylebrandt committed:
test empty via TimeSeriesReaderFromFrames

kylebrandt committed:
allow additional frames in empty case

kylebrandt committed:
some linting

kylebrandt committed:
more lint

kylebrandt committed:
move types into own packages

kylebrandt committed:
test file of dir listing

kylebrandt committed:
start reworking numeric

kylebrandt committed:
quick basic example for multi series

kylebrandt committed:
more updating of numeric, still wip

kylebrandt committed:
update New pattern for numeric, still need to update getRefs/validate for numeric

kylebrandt committed:
misc comments

ying-jeanne commented on experimental/sdata/timeseries/multi.go:
```diff
@@ -0,0 +1,227 @@
+}
+
+// values must be a numeric slice such as []int64, []float64, []*float64, etc or []bool / []*bool.
+func (mfs *MultiFrame) AddSeries(metricName string, l data.Labels, t []time.Time, values interface{}) error {
```
here []time.Time should be []*time.Time, since we can miss data for certain data point, so the time would be able to be nil.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/grafana/grafana-plugin-sdk-go?pullRequest=440) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/grafana/grafana-plugin-sdk-go?pullRequest=440) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/grafana/grafana-plugin-sdk-go?pullRequest=440) it.</sub>

kylebrandt committed:
Merge branch 'main' into dataSchemaPlay

kylebrandt committed:
Merge branch 'main' into dataSchemaPlay

kylebrandt committed:
some of lint

kylebrandt committed:
remove validate from series reader interface

kylebrandt committed:
remove validate from numeric reader interface

kylebrandt committed:
remove validate from series readers

kylebrandt committed:
remove getRefs from numeric readers

papagian reviewed (approved):
:trophy:

kylebrandt committed:
review fixes
